File: src/c/ownership.c
// 1️⃣ Ownership and Automatic Release vs C Manual Memory Management
// 
// Compile: gcc -O2 -S src/c/ownership.c -o target/ownership.s
// Assembly file: target/ownership.s
//
// Key assembly differences to observe:
// - Explicit malloc call for memory allocation
// - Explicit free call for memory deallocation
// - Forgetting free() causes memory leaks
// - sprintf has potential buffer overflow risks (safe in this example)
// - No compile-time memory safety guarantees
// - Risk of use-after-free if accessing s after free()
// - Risk of double-free if free() called twice
// - Manual memory management overhead and error-prone nature

#include <stdio.h>
#include <stdlib.h>

int main() {
    char *s = malloc(6);
    sprintf(s, "hello");
    printf("%s\n", s);
    free(s);  // Manual deallocation required - assembly shows free call
    return 0;
} 

File: src/rust/ownership.rs
// 1️⃣ Ownership and Automatic Release vs C Manual Memory Management
// 
// Compile: cargo rustc --bin ownership --release -- --emit=asm
// Assembly file: target/release/deps/ownership-*.s
//
// Key assembly differences to observe:
// - Rust automatically inserts drop calls at function end
// - String::from triggers heap allocation function calls
// - Scope termination automatically calls drop_in_place for memory cleanup
// - Release mode may optimize away some intermediate steps
// - Memory safety guaranteed at compile time through ownership rules
// - No risk of use-after-free or double-free bugs
// - Zero-cost abstraction: ownership tracking happens at compile time

fn main() {
    let s = String::from("hello");
    println!("{}", s);
} // s automatically dropped here - assembly shows drop_in_place call 